dynamiq/memory/memory.py

```diff
@@ -0,0 +1,91 @@
+from datetime import datetime
+
+from dynamiq.memory.backend import InMemory, MemoryBackend
+from dynamiq.memory.config import Config
+from dynamiq.prompts import Message, MessageRole
+from dynamiq.utils.logger import logger
+
+
+class Memory:
+    """Manages the storage and retrieval of messages."""
+
+    def __init__(self, config: Config = Config(), backend: MemoryBackend = InMemory()):
+        """Initializes the Memory with the given configuration and backend.
+
+        If no backend is provided, an InMemory backend is used by default.
+        """
+        if not isinstance(backend, MemoryBackend):
+            raise TypeError("backend must be an instance of Backend")
+        self.config = config
+        self.backend = backend
+
+    def add(self, role: MessageRole, content: str, metadata: dict = None):
+        """Adds a message to the memory."""
```

Comment:
I prefer more extensive docstrings with detailed information about input parameters, which can later be rendered with mkdocs. For instance, what parameters are expected in metadata, etc.
However, this is a minor preference on my part, and it can remain as it is now.